master/pkg/searcher/asha.go

```diff
@@ -1,169 +1,181 @@
+// AsyncHalvingSearch implements a search using the asynchronous successive halving algorithm
+// (ASHA). The experiment will run until the target number of trials have been completed
+// in the bottom rung and no further promotions can be made to higher rungs.
 type asyncHalvingSearch struct {
 	defaultSearchMethod
 	model.AsyncHalvingConfig
 
 	rungs      []*rung
 	trialRungs map[RequestID]int
 	// earlyExitTrials contains trials that exited early that are still considered in the search.
-	earlyExitTrials   map[RequestID]bool
-	expectedWorkloads int
-	trialsCompleted   int
+	earlyExitTrials     map[RequestID]bool
+	totalStepsCompleted int
+	maxTrials           int
+	trialsCompleted     int
 }
 
 const ashaExitedMetricValue = math.MaxFloat64
 
 func newAsyncHalvingSearch(config model.AsyncHalvingConfig) SearchMethod {
 	rungs := make([]*rung, 0, config.NumRungs)
-	expectedSteps := 0
 	for id := 0; id < config.NumRungs; id++ {
 		compound := math.Pow(config.Divisor, float64(config.NumRungs-id-1))
 		stepsNeeded := max(int(float64(config.TargetTrialSteps)/compound), 1)
-		startTrials := max(int(compound), 1)
+		startTrials := max(min(int(compound), config.MaxTrials), 1)
 		rungs = append(rungs,
 			&rung{
-				stepsNeeded: stepsNeeded,
-				startTrials: startTrials,
-			},
-		)
-		if id == 0 {
-			expectedSteps += stepsNeeded * startTrials
-		} else {
-			expectedSteps += (stepsNeeded - rungs[id-1].stepsNeeded) * startTrials
-		}
-	}
-
-	expectedWorkloads := 0
-	multiplier := float64(config.StepBudget) / float64(expectedSteps)
-	for id := 0; id < config.NumRungs; id++ {
-		cur := rungs[id]
-		cur.startTrials = int(multiplier * float64(cur.startTrials))
-		if id != 0 {
-			prev := rungs[id-1]
-			cur.stepsNeeded = max(cur.stepsNeeded, prev.stepsNeeded+1)
-			cur.startTrials = max(min(cur.startTrials, prev.startTrials), 1)
-			prev.promoteTrials = cur.startTrials
-			expectedWorkloads += (cur.stepsNeeded - prev.stepsNeeded + 1) * cur.startTrials
-		} else {
-			expectedWorkloads += (cur.stepsNeeded + 1) * cur.startTrials
-		}
+				divisor:           config.Divisor,
+				stepsNeeded:       stepsNeeded,
+				startTrials:       startTrials,
+				outstandingTrials: 0})
 	}
 
 	return &asyncHalvingSearch{
 		AsyncHalvingConfig: config,
 		rungs:              rungs,
 		trialRungs:         make(map[RequestID]int),
 		earlyExitTrials:    make(map[RequestID]bool),
-		expectedWorkloads:  expectedWorkloads,
+		maxTrials:          config.MaxTrials,
 	}
 }
 
 type trialMetric struct {
 	requestID RequestID
 	metric    float64
+	promoted  bool
+	closed    bool
 }
 
 // rung describes a set of trials that are to be trained for the same number of steps.
 type rung struct {
-	stepsNeeded   int
-	metrics       []trialMetric
-	startTrials   int
-	promoteTrials int
+	divisor           float64
+	stepsNeeded       int
+	metrics           []trialMetric
+	startTrials       int
+	outstandingTrials int
```

Comment:
question: do we need this? Shouldn't `outstandingTrials == startTrials - len(metrics)`?

Reply:
startTrials is not a counter and is an int used for initialOperations.  len(metrics) is completed trials in a rung but not pending so that's why outstandingTrials is needed.